File: java-examples/scenario2-data-pipeline/stream-optimization/before/DataProcessingService.java
package com.enterprise.optimization.service;

import java.util.List;
import java.util.ArrayList;
import java.util.Map;
import java.util.HashMap;
import java.util.stream.Collectors;
import java.time.LocalDate;
import java.math.BigDecimal;

/**
 * Service for data processing operations using Java streams.
 * This implementation has several performance issues.
 */
public class DataProcessingService {

    /**
     * Process transaction data to calculate total amount by category.
     * Inefficient implementation using multiple stream operations.
     */
    public Map<String, BigDecimal> calculateTotalByCategory(List<Transaction> transactions) {
        Map<String, BigDecimal> result = new HashMap<>();
        
        // Filter valid transactions
        List<Transaction> validTransactions = transactions.stream()
            .filter(t -> t.getAmount().compareTo(BigDecimal.ZERO) > 0)
            .collect(Collectors.toList());
        
        // Get all categories
        List<String> categories = validTransactions.stream()
            .map(Transaction::getCategory)
            .distinct()
            .collect(Collectors.toList());
        
        // Calculate total for each category
        for (String category : categories) {
            BigDecimal total = validTransactions.stream()
                .filter(t -> t.getCategory().equals(category))
                .map(Transaction::getAmount)
                .reduce(BigDecimal.ZERO, BigDecimal::add);
            
            result.put(category, total);
        }
        
        return result;
    }
    
    /**
     * Find transactions that meet specific criteria.
     * Inefficient implementation with repeated streams.
     */
    public List<Transaction> findTransactionsByCriteria(List<Transaction> transactions, 
                                                       String merchant, 
                                                       LocalDate startDate, 
                                                       LocalDate endDate, 
                                                       BigDecimal minAmount) {
        // Filter by merchant
        List<Transaction> byMerchant = transactions.stream()
            .filter(t -> t.getMerchant().equalsIgnoreCase(merchant))
            .collect(Collectors.toList());
        
        // Filter by date range
        List<Transaction> byDateRange = byMerchant.stream()
            .filter(t -> !t.getDate().isBefore(startDate) && !t.getDate().isAfter(endDate))
            .collect(Collectors.toList());
        
        // Filter by minimum amount
        List<Transaction> result = byDateRange.stream()
            .filter(t -> t.getAmount().compareTo(minAmount) >= 0)
            .collect(Collectors.toList());
        
        return result;
    }
    
    /**
     * Group transactions by month and calculate monthly statistics.
     * Inefficient implementation with multiple iterations.
     */
    public List<MonthlyStatistics> calculateMonthlyStatistics(List<Transaction> transactions) {
        List<MonthlyStatistics> result = new ArrayList<>();
        
        // Get all year-months
        List<YearMonth> yearMonths = transactions.stream()
            .map(t -> new YearMonth(t.getDate().getYear(), t.getDate().getMonthValue()))
            .distinct()
            .collect(Collectors.toList());
        
        // Calculate statistics for each month
        for (YearMonth yearMonth : yearMonths) {
            List<Transaction> monthTransactions = transactions.stream()
                .filter(t -> t.getDate().getYear() == yearMonth.year && t.getDate().getMonthValue() == yearMonth.month)
                .collect(Collectors.toList());
            
            // Calculate total amount
            BigDecimal totalAmount = monthTransactions.stream()
                .map(Transaction::getAmount)
                .reduce(BigDecimal.ZERO, BigDecimal::add);
            
            // Calculate average amount
            BigDecimal avgAmount = totalAmount.divide(BigDecimal.valueOf(monthTransactions.size()), 2, BigDecimal.ROUND_HALF_UP);
            
            // Find largest transaction
            Transaction largestTransaction = monthTransactions.stream()
                .max((t1, t2) -> t1.getAmount().compareTo(t2.getAmount()))
                .orElse(null);
            
            // Create monthly statistics
            MonthlyStatistics stats = new MonthlyStatistics();
            stats.yearMonth = yearMonth;
            stats.transactionCount = monthTransactions.size();
            stats.totalAmount = totalAmount;
            stats.averageAmount = avgAmount;
            stats.largestTransactionAmount = largestTransaction != null ? largestTransaction.getAmount() : BigDecimal.ZERO;
            
            result.add(stats);
        }
        
        return result;
    }
    
    /**
     * Transaction class for the example
     */
    public static class Transaction {
        private String id;
        private LocalDate date;
        private BigDecimal amount;
        private String merchant;
        private String category;
        
        // Getters and setters
        public String getId() { return id; }
        public void setId(String id) { this.id = id; }
        
        public LocalDate getDate() { return date; }
        public void setDate(LocalDate date) { this.date = date; }
        
        public BigDecimal getAmount() { return amount; }
        public void setAmount(BigDecimal amount) { this.amount = amount; }
        
        public String getMerchant() { return merchant; }
        public void setMerchant(String merchant) { this.merchant = merchant; }
        
        public String getCategory() { return category; }
        public void setCategory(String category) { this.category = category; }
    }
    
    /**
     * Year-Month helper class
     */
    public static class YearMonth {
        public int year;
        public int month;
        
        public YearMonth(int year, int month) {
            this.year = year;
            this.month = month;
        }
        
        @Override
        public boolean equals(Object obj) {
            if (!(obj instanceof YearMonth)) return false;
            YearMonth other = (YearMonth) obj;
            return this.year == other.year && this.month == other.month;
        }
        
        @Override
        public int hashCode() {
            return 31 * year + month;
        }
    }
    
    /**
     * Monthly statistics class
     */
    public static class MonthlyStatistics {
        public YearMonth yearMonth;
        public int transactionCount;
        public BigDecimal totalAmount;
        public BigDecimal averageAmount;
        public BigDecimal largestTransactionAmount;
    }
} 

File: dotnet-examples/scenario2-data-pipeline/linq-optimization/before/DataProcessingService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Microsoft.Extensions.Logging;

namespace Enterprise.DataProcessing.Services
{
    /// <summary>
    /// Service for data processing operations using LINQ.
    /// This implementation has several performance issues.
    /// </summary>
    public class DataProcessingService
    {
        private readonly ILogger<DataProcessingService> _logger;

        public DataProcessingService(ILogger<DataProcessingService> logger)
        {
            _logger = logger;
        }

        /// <summary>
        /// Process transaction data to calculate total amount by category.
        /// Inefficient implementation using multiple LINQ operations.
        /// </summary>
        public Dictionary<string, decimal> CalculateTotalByCategory(List<Transaction> transactions)
        {
            var result = new Dictionary<string, decimal>();
            
            // Filter valid transactions
            var validTransactions = transactions
                .Where(t => t.Amount > 0)
                .ToList();
            
            // Get all categories
            var categories = validTransactions
                .Select(t => t.Category)
                .Distinct()
                .ToList();
            
            // Calculate total for each category
            foreach (var category in categories)
            {
                var total = validTransactions
                    .Where(t => t.Category == category)
                    .Sum(t => t.Amount);
                
                result.Add(category, total);
            }
            
            return result;
        }
        
        /// <summary>
        /// Find transactions that meet specific criteria.
        /// Inefficient implementation with repeated LINQ queries.
        /// </summary>
        public List<Transaction> FindTransactionsByCriteria(List<Transaction> transactions, 
                                                          string merchant, 
                                                          DateTime startDate, 
                                                          DateTime endDate, 
                                                          decimal minAmount)
        {
            // Filter by merchant
            var byMerchant = transactions
                .Where(t => t.Merchant.Equals(merchant, StringComparison.OrdinalIgnoreCase))
                .ToList();
            
            // Filter by date range
            var byDateRange = byMerchant
                .Where(t => t.Date >= startDate && t.Date <= endDate)
                .ToList();
            
            // Filter by minimum amount
            var result = byDateRange
                .Where(t => t.Amount >= minAmount)
                .ToList();
            
            return result;
        }
        
        /// <summary>
        /// Group transactions by month and calculate monthly statistics.
        /// Inefficient implementation with multiple iterations.
        /// </summary>
        public List<MonthlyStatistics> CalculateMonthlyStatistics(List<Transaction> transactions)
        {
            var result = new List<MonthlyStatistics>();
            
            // Get all year-months
            var yearMonths = transactions
                .Select(t => new { Year = t.Date.Year, Month = t.Date.Month })
                .Distinct()
                .ToList();
            
            // Calculate statistics for each month
            foreach (var yearMonth in yearMonths)
            {
                var monthTransactions = transactions
                    .Where(t => t.Date.Year == yearMonth.Year && t.Date.Month == yearMonth.Month)
                    .ToList();
                
                // Calculate total amount
                var totalAmount = monthTransactions.Sum(t => t.Amount);
                
                // Calculate average amount
                var avgAmount = monthTransactions.Average(t => t.Amount);
                
                // Find largest transaction
                var largestTransaction = monthTransactions
                    .OrderByDescending(t => t.Amount)
                    .FirstOrDefault();
                
                // Create monthly statistics
                var stats = new MonthlyStatistics
                {
                    Year = yearMonth.Year,
                    Month = yearMonth.Month,
                    TransactionCount = monthTransactions.Count,
                    TotalAmount = totalAmount,
                    AverageAmount = avgAmount,
                    LargestTransactionAmount = largestTransaction?.Amount ?? 0
                };
                
                result.Add(stats);
            }
            
            return result;
        }
        
        /// <summary>
        /// Process a large list of orders.
        /// Inefficient implementation with multiple expensive operations.
        /// </summary>
        public List<OrderSummary> ProcessOrders(List<Order> orders)
        {
            // Process each order and create a summary
            var orderSummaries = new List<OrderSummary>();
            
            foreach (var order in orders)
            {
                // Get items for this order
                var orderItems = order.Items;
                
                // Calculate total price
                decimal totalPrice = 0;
                foreach (var item in orderItems)
                {
                    totalPrice += item.Price * item.Quantity;
                }
                
                // Calculate discount
                decimal discount = 0;
                if (totalPrice > 1000)
                {
                    discount = totalPrice * 0.1m; // 10% discount for orders over $1000
                }
                else if (totalPrice > 500)
                {
                    discount = totalPrice * 0.05m; // 5% discount for orders over $500
                }
                
                // Count items by category
                var categoryCounts = new Dictionary<string, int>();
                foreach (var item in orderItems)
                {
                    if (categoryCounts.ContainsKey(item.Category))
                    {
                        categoryCounts[item.Category] += item.Quantity;
                    }
                    else
                    {
                        categoryCounts.Add(item.Category, item.Quantity);
                    }
                }
                
                // Get top category
                string topCategory = null;
                int maxCount = 0;
                foreach (var kvp in categoryCounts)
                {
                    if (kvp.Value > maxCount)
                    {
                        maxCount = kvp.Value;
                        topCategory = kvp.Key;
                    }
                }
                
                // Create and add summary
                var summary = new OrderSummary
                {
                    OrderId = order.OrderId,
                    CustomerId = order.CustomerId,
                    OrderDate = order.OrderDate,
                    ItemCount = orderItems.Sum(i => i.Quantity),
                    TotalAmount = totalPrice,
                    Discount = discount,
                    NetAmount = totalPrice - discount,
                    TopCategory = topCategory
                };
                
                orderSummaries.Add(summary);
            }
            
            return orderSummaries;
        }
    }

    /// <summary>
    /// Transaction class for the example
    /// </summary>
    public class Transaction
    {
        public string Id { get; set; }
        public DateTime Date { get; set; }
        public decimal Amount { get; set; }
        public string Merchant { get; set; }
        public string Category { get; set; }
    }
    
    /// <summary>
    /// Monthly statistics class
    /// </summary>
    public class MonthlyStatistics
    {
        public int Year { get; set; }
        public int Month { get; set; }
        public int TransactionCount { get; set; }
        public decimal TotalAmount { get; set; }
        public decimal AverageAmount { get; set; }
        public decimal LargestTransactionAmount { get; set; }
    }
    
    /// <summary>
    /// Order class for the example
    /// </summary>
    public class Order
    {
        public string OrderId { get; set; }
        public string CustomerId { get; set; }
        public DateTime OrderDate { get; set; }
        public List<OrderItem> Items { get; set; } = new List<OrderItem>();
    }
    
    /// <summary>
    /// Order item class
    /// </summary>
    public class OrderItem
    {
        public string ItemId { get; set; }
        public string Name { get; set; }
        public string Category { get; set; }
        public decimal Price { get; set; }
        public int Quantity { get; set; }
    }
    
    /// <summary>
    /// Order summary class
    /// </summary>
    public class OrderSummary
    {
        public string OrderId { get; set; }
        public string CustomerId { get; set; }
        public DateTime OrderDate { get; set; }
        public int ItemCount { get; set; }
        public decimal TotalAmount { get; set; }
        public decimal Discount { get; set; }
        public decimal NetAmount { get; set; }
        public string TopCategory { get; set; }
    }
} 